lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,260 @@
+                sink << datagram_builders[no_prefix].distribution_value_packed(
+                  agg_value.name,
+                  agg_value.type.to_s,
+                  agg_value.value,
+                  CONST_SAMPLE_RATE,
+                  agg_value.tags,
+                )
+              when GAUGE
+                sink << datagram_builders[no_prefix].g(
+                  agg_value.name,
+                  agg_value.value,
+                  CONST_SAMPLE_RATE,
+                  agg_value.tags,
+                )
+              else
+                StatsD.logger.error { "[#{self.class.name}] Unknown aggregation type: #{agg_value.type}" }
+              end
+            end
+          end
+        end
+      end
+
+      # @param sink [#<<] The sink to write the aggregated metrics to.
+      # @param datagram_builder_class [Class] The class to use for building datagrams.
+      # @param prefix [String] The prefix to add to all metrics.
+      # @param default_tags [Array<String>] The tags to add to all metrics.
+      # @param flush_interval [Float] The interval at which to flush the aggregated metrics.
+      # @param max_values [Integer] The maximum number of values to aggregate before flushing.
+      def initialize(sink, datagram_builder_class, prefix, default_tags, flush_interval: 5.0, max_values: 100)
+        @sink = sink
+        @datagram_builder_class = datagram_builder_class
+        @metric_prefix = prefix
+        @default_tags = default_tags
+        @datagram_builders = {
+          true: nil,
+          false: nil,
+        }
+        @max_values = max_values
+
+        # Mutex protects the aggregation_state and flush_thread from concurrent access
+        @mutex = Mutex.new
+        @aggregation_state = {}
+
+        @pid = Process.pid
+        @flush_interval = flush_interval
+        @flush_thread = Thread.new do
+          Thread.current.abort_on_exception = true
+          loop do
+            sleep(@flush_interval)
+            thread_healthcheck
+            flush
+          end
+        end
+
+        ObjectSpace.define_finalizer(
+          self,
+          self.class.finalize(@aggregation_state, @sink, @datagram_builders, @datagram_builder_class, @default_tags),
+        )
+      end
+
+      # Increment a counter by a given value and save it for later flushing.
+      # @param name [String] The name of the counter.
+      # @param value [Integer] The value to increment the counter by.
+      # @param tags [Hash{String, Symbol => String},Array<String>] The tags to attach to the counter.
+      # @param no_prefix [Boolean] If true, the metric will not be prefixed.
+      # @return [void]
+      def increment(name, value = 1, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).c(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
```

Comment:
This sorting and tag serialization is still duplicated work as we'll need to do it again when building the datagram.

Reply:
Reworked this part, kept only type, no_prefix to avoid type conversions (string to symbol, string to bool) and `value`, obviously.